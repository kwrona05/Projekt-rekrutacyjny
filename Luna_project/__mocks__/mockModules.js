export const mockModules = [
    {
        id: "0a0f77eb1-50a0-4d98-8116-064fc5a84693",
        name: "Hydroponic module 1",
        description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Eget nullam non nisi est sit amet.",
        available: true,
        targetTemperature: 10.0,
    },
    {
        id: "4d0aa62c-b1a9-489d-b4a2-fc16b878ba47",
        name: "Hydroponic module 2",
        description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nulla aliquet enim tortor at auctor.",
        available: false,
        targetTemperature: 15.5,
    },
    {
        id: "d4928094-8ef8-48be-823a-4cddef643249",
        name: "Hydroponic module 3",
        description:
            "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Diam in arcu cursus euismod quis viverra nibh cras.",
        available: true,
        targetTemperature: 20.0,
    },
  ]
  