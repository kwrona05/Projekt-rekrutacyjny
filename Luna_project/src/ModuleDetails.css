* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  font-family: 'Poppins', sans-serif;
}

body {
  background-color: #f4f4f4;
  color: #333;
  line-height: 1.6;
}

.details-container {
    max-width: 900px;
    margin: 40px auto;
    padding: 20px;
    background-color: white;
    border-radius: 10px;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  }
  
  h2.name {
    font-size: 2.5rem;
    color: #007BFF;
    margin-bottom: 20px;
  }

  h3 {
    color: #007BFF;
  }
  
  p {
    font-size: 1.2rem;
    margin-bottom: 10px;
  }
  
  strong {
    color: #007BFF;
  }
  
  button {
    background-color: #007BFF;
    color: white;
    border: none;
    padding: 12px 20px;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 5px;
    transition: background-color 0.3s ease;
    margin-right: 10px;
  }
  
  button:hover {
    background-color: #0056b3;
  }
  
  .edit-btn:disabled {
    background-color: #ccc;
    cursor: not-allowed;
  }
  
  .back-btn {
    background-color: #333;
  }

.module-history {
    font-size: 1.8rem;
    margin-top: 40px;
    margin-bottom: 20px;
    color: #007BFF;
  }
  
  .recharts-wrapper {
    margin-top: 20px;
  }

.error {
    color: red;
    font-size: 0.9rem;
  }

.modal {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background: rgba(0, 0, 0, 0.5);
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 1000;
  }
  
  .modal-content {
    background-color: white;
    padding: 30px;
    border-radius: 10px;
    max-width: 500px;
    width: 100%;
  }
  
  .modal-content h3 {
    font-size: 2rem;
    color: #007BFF;
    margin-bottom: 20px;
  }
  
  .modal-content label {
    display: block;
    font-size: 1.1rem;
    margin-bottom: 5px;
  }
  
  .modal-content input {
    width: 100%;
    padding: 10px;
    margin-bottom: 15px;
    border: 1px solid #ccc;
    border-radius: 5px;
  }
  
  .modal-btn {
    background-color: #007BFF;
    margin-right: 10px;
  }
  
  .modal-btn:hover {
    background-color: #0056b3;
  }

form {
    margin: 20px 0;
    display: flex;
    flex-direction: column;
  }
  
  form label {
    font-size: 1.1rem;
    margin-bottom: 10px;
  }
  
  form input,
  form select {
    padding: 10px;
    border-radius: 5px;
    border: 1px solid #ccc;
    margin-bottom: 20px;
    font-size: 1rem;
  }
  
  .fetch-history {
    align-self: center;
  }
  
  input[type="datetime-local"] {
    background-color: #fff;
    border: 1px solid #ccc;
  }
  
  .fetch-history {
    background-color: #28a745;
  }
  
  .fetch-history:hover {
    background-color: #218838;
  }

  .available-text {
    color: #28a745; /* Zielony */
  }
  
  .unavailable-text {
    color: #dc3545; /* Czerwony */
  }
