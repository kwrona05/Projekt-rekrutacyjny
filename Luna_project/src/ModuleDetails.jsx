import { useEffect, useState } from "react";
import { useParams, useNavigate } from "react-router-dom";
import "./App.css";
import ModuleTemperature from "./ModuleTemperature";
import {
  LineChart,
  Line,
  ResponsiveContainer,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
} from "recharts";

function ModuleDetails() {
  const { id } = useParams();
  const [module, setModule] = useState(null);
  const [error, setError] = useState({ isError: false, error: null });
  const [isEditModal, setIsEditModal] = useState(false);
  const [historyData, setHistoryData] = useState();
  const [editData, setEditData] = useState({
    name: "",
    description: "",
    targetTemperature: "",
  });
  const [validation, setValidation] = useState({});
  const [dataRange, setDataRange] = useState({
    start: "",
    stop: "",
  });
  const [mode, setMode] = useState("hourly");
  const navigate = useNavigate();

  useEffect(() => {
    const fetchModuleDetails = async () => {
      try {
        const response = await fetch(`http://localhost:3001/modules/${id}`);
        if (response.status === 404) {
          throw new Error("Page not found");
        }
        console.log(response);
        const data = await response.json();
        setModule(data);
      } catch (error) {
        console.error("Error fetching module details:", error);
        setError({ error, isError: true });
      }
    };

    fetchModuleDetails();
  }, [id]);

  const fetchHistoryData = async () => {
    try {
      const response = await fetch(
        `http://localhost:3001/modules/${id}/history?start=${dateRange.start}&stop=${dateRange.stop}&mode=${mode}`
      );
      if (!response.ok) {
        throw new Error("Something goes wrong during fetching history data");
      }
      const data = await response.json();
      setHistoryData(data);
    } catch (error) {
      console.error(
        "Something goes wrong during fetching history data:",
        error
      );
      setError({ error, isError: true });
    }
  };

  const handleDataChange = (e) => {
    const { name, value } = e.target;
    setDataRange((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  const handleModeChange = (e) => {
    setMode(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    fetchHistoryData();
  };

  if (error.isError) {
    return (
      <div>
        <h3>Something went wrong...</h3>
        {error?.error?.message && <h4>{error.error.message}</h4>}
      </div>
    );
  }

  const handleOpenModal = () => {
    setEditData({
      name: module.name,
      description: module.description,
      targetTemperature: module.targetTemperature,
    });
    setIsEditModal(true);
  };

  const handleCloseModal = () => {
    setIsEditModal(false);
    setValidation({});
  };

  const handleEditChange = (e) => {
    const { name, value } = e.target;
    setEditData((prevState) => ({ ...prevState, [name]: value }));
  };

  const validateForm = () => {
    const errors = {};
    if (!editData.name.trim()) errors.name = "Name is required";
    if (!editData.description.trim())
      errors.description = "Description is required";
    if (
      isNaN(editData.targetTemperature) ||
      editData.targetTemperature < 0 ||
      editData.targetTemperature > 40
    ) {
      errors.targetTemperature = "Temperature has to be between 0 and 40";
    }
    return errors;
  };

  const handleSaveChanges = async () => {
    const validationErrors = validateForm();
    if (Object.keys(validationErrors).length > 0) {
      setValidation(validationErrors);
      return;
    }

    try {
      const response = await fetch(`http://localhost:3001/modules/${id}`, {
        method: "PATCH",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(editData),
      });

      if (!response.ok) {
        throw new Error("Update module details failed");
      }

      const updateModule = await response.json();
      setModule(updateModule);
      handleCloseModal();
    } catch (error) {
      console.error("Error during updating module details:", error);
      setError({ error, isError: true });
    }
  };

  if (error.isError) {
    return (
      <div>
        <h3>Something went wrong...</h3>
        {error?.error?.message && <h4>{error.error.message}</h4>}
      </div>
    );
  }

  if (!module) {
    return <div>Loading details...</div>;
  }

  return (
    <div>
      <h2>{module.name}</h2>
      <p>{module.description}</p>
      <p>
        <strong>Available:</strong> {module.available ? "Yes" : "No"}
      </p>
      <p>
        <strong>Target Temperature:</strong> {module.targetTemperature} °C
      </p>
      <ModuleTemperature targetTemperature={module.targetTemperature} />

      {module.available ? (
        <button onClick={handleOpenModal}>Edit module details</button>
      ) : (
        <button disabled>Module unavailable</button>
      )}

      <button onClick={() => navigate("/")}>Back to Module List</button>

      {/*History*/}
      <h3>Module Temperature History</h3>
      <form onSubmit={handleSubmit}>
        <label>
          Start Date:
          <input
            type="datetime-local"
            name="start"
            value={dataRange.start}
            onChange={handleDataChange}
            required
          />
        </label>
        <label>
          Stop Date:
          <input
            type="datetime-local"
            name="stop"
            value={dataRange.stop}
            onChange={handleDataChange}
            required
          />
        </label>
        <label>
          Mode:
          <select value={mode} onChange={handleModeChange}>
            <option value="hourly">Hourly</option>
            <option value="daily">Daily</option>
          </select>
        </label>
        <button type="submit">Fetch History</button>
      </form>

      {/*Chart*/}
      {historyData.length > 0 && (
        <ResponsiveContainer width="100%" height={400}>
          <LineChart data={historyData}>
            <CartesianGrid strokeDasharray="3 3" />
            <XAxis
              dataKey="timestamp"
              tickFormatter={(tick) => new Date(tick).toLocaleString()}
            />
            <YAxis />
            <Tooltip
              labelFormatter={(label) => new Date(label).toLocaleString()}
            />
            <Legend />
            <Line
              type="monotone"
              dataKey="temperature"
              stroke="#8884d8"
              activeDot={{ r: 8 }}
            />
          </LineChart>
        </ResponsiveContainer>
      )}

      {/* Modal */}
      {isEditModal && (
        <div className="modal">
          <div className="modal-content">
            <h3>Edit Module</h3>
            <label>
              Name:
              <input
                type="text"
                name="name"
                value={editData.name}
                onChange={handleEditChange}
              />
              {validation.name && (
                <span className="error">{validation.name}</span>
              )}
            </label>
            <label>
              Description:
              <input
                type="text"
                name="description"
                value={editData.description}
                onChange={handleEditChange}
              />
              {validation.description && (
                <span className="error">{validation.description}</span>
              )}
            </label>
            <label>
              Target Temperature (°C):
              <input
                type="number"
                name="targetTemperature"
                value={editData.targetTemperature}
                onChange={handleEditChange}
              />
              {validation.targetTemperature && (
                <span className="error">{validation.targetTemperature}</span>
              )}
            </label>
            <button onClick={handleSaveChanges}>Save changes</button>
            <button onClick={handleCloseModal}>Cancel</button>
          </div>
        </div>
      )}
    </div>
  );
}
export default ModuleDetails;
