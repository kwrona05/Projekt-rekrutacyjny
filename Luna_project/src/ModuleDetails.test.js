/* eslint-disable no-undef */
import "@testing-library/jest-dom";

import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import ModuleDetails from "./ModuleDetails";
import { mockModules } from "../__mocks__/mockModules";

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(mockModules[0]),
  })
);

const mockModuleWithTemperature = { ...mockModules[0], temperature: 11 };

beforeEach(() => {
  fetch.mockClear();
});

describe("COMPONENT TESTS", () => {
  test("Is button in document", async () => {
    render(
      <MemoryRouter>
        <ModuleDetails modules={[mockModuleWithTemperature]} />
      </MemoryRouter>
    );

    await waitFor(() => {
      const button = screen.getByTestId("edit-module");
      expect(button).not.toBeUndefined();
      expect(button).toHaveTextContent("Edit module");
    });
  });

  test("testing text in h3 element", async () => {
    render(
      <MemoryRouter>
        <ModuleDetails modules={[mockModuleWithTemperature]} />
      </MemoryRouter>
    );

    await waitFor(() => {
      const h3 = screen.getByTestId("history-module");
      expect(h3).not.toBeUndefined();
      expect(h3).toHaveTextContent("Module Temperature History");
    });
  });

  test("Are two inputs in document", async () => {
    render(
      <MemoryRouter>
        <ModuleDetails modules={[mockModuleWithTemperature]} />
      </MemoryRouter>
    );

    await waitFor(() => {
      const input = screen.getAllByTestId("input-element");
      expect(input).not.toBeUndefined();
      expect(input.length).toBe(2);
    });
  });

  test("Is select defined", async () => {
    render(
      <MemoryRouter>
        <ModuleDetails modules={[mockModuleWithTemperature]} />
      </MemoryRouter>
    );

    await waitFor(() => {
      const select = screen.getByTestId("select-element");
      expect(select).not.toBeUndefined();
    });
  });

  test("Are two opitons with correct text in document", async () => {
    render(
      <MemoryRouter>
        <ModuleDetails modules={[mockModuleWithTemperature]} />
      </MemoryRouter>
    );

    await waitFor(() => {
      const options = screen.getAllByRole("option");
      expect(options).not.toBeUndefined();
      expect(options.length).toBe(2);

      const hourly = screen.getByTestId("hourly");
      expect(hourly).not.toBeUndefined();
      expect(hourly).toHaveTextContent("Hourly");

      const daily = screen.getByTestId("daily");
      expect(daily).not.toBeUndefined();
      expect(daily).toHaveTextContent("Daily");
    });
  });
});
describe("SNAPSHOT TESTS", () => {
  test("It matches saved snapshot", async () => {
    jest.mock("react-router-dom", () => ({
      ...jest.requireActual("react-router-dom"),
      useParams: jest.fn(() => mockModules[0].id),
    }));

    render(
      <MemoryRouter>
        <ModuleDetails modules={[mockModuleWithTemperature]} />
      </MemoryRouter>
    );

    await waitFor(() => {
      const button = screen.getByTestId("edit-module");
      expect(button).toMatchSnapshot();
    });
  });
});
