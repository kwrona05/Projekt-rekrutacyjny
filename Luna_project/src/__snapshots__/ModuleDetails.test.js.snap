// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT TESTS It matches saved snapshot 1`] = `
<button
  class="edit-btn"
  data-testid="edit-module"
>
  Edit module details
</button>
`;
